api/placer.py

```diff
@@ -138,7 +138,8 @@ def check(self):
         metadata_validator = validators.from_schema_path(payload_schema_uri)
         metadata_validator(self.metadata, 'POST')
 
-        targets = self.create_hierarchy(self.metadata)
+        # If not a superuser request, pass uid of user making the upload request
```

Comment:
Yeah, that's a good idea. 
You could add a "get_target_object" method to each hierarchy uploader (in a similar way to how create_hierarchy is defined for them), and then compute permissions on it.
It could be a little convoluted but you will extract some complexity from the hierarchy creation/update part.